### PR TITLE
Add script to cover npiv hostdev passthrough cases

### PR DESCRIPTION
**Pull request by lento-sun**
1. hot attach hot detach
2. cold attach hot detach

Signoff: Yi Sun (lentosun@163.com)

depend on avocado-framework/avocado-vt#797
https://github.com/avocado-framework/avocado-vt/pull/826

### CODE REVIEW

**Comment by lento-sun** [comment]
@harish-24 
Could you please help to check if this script can be passed in your env? thx

**Comment by lento-sun** [comment]
Test result. (x86_64 machine)
kernel-3.10.0-514.el7.x86_64
qemu-kvm-rhev-2.6.0-29.el7.x86_64
libvirt-2.0.0-10.el7_3.2.x86_64


[root@bootp-73-75-161 npiv]# avocado run --vt-type libvirt type_specific.lento.npiv.hostdev_passthrough.positive_testing
JOB ID     : 9accefe5c2a4befbd18697411c3dd63220c3040c
JOB LOG    : /root/avocado/job-results/job-2016-12-01T18.37-9accefe/job.log
TESTS      : 2
 (1/2) type_specific.lento.npiv.hostdev_passthrough.positive_testing.hot_attach_hot_detach_hostdev: PASS (21.54 s)
 (2/2) type_specific.lento.npiv.hostdev_passthrough.positive_testing.cold_attach_hot_detach_hostdev: PASS (37.84 s)

**Comment by harish-24** [comment]
@lento-sun Please mention the dependency PR here. https://github.com/avocado-framework/avocado-vt/pull/797

File: libvirt/tests/src/npiv/npiv_hostdev_passthrough.py [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+        if attach_method == "hot":
+            #attach-device the lun's hostdev xml to guest vm
+            result = virsh.attach_device(vm_name, new_hostdev_xml.xml)
+            status = result.exit_status
+            if status_error == "yes":
+                if status:
+                    raise exceptions.TestFail("Attach device failed: %s",
+                                              result.stderr)
+        elif attach_method == "cold":
+            # TODO
```
Comment (lento-sun) [inline libvirt/tests/src/npiv/npiv_hostdev_passthrough.py]: redundant TODO, should remove

File: libvirt/tests/src/npiv/npiv_hostdev_passthrough.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+        #find first online hba
+        online_hbas = []
+        vhba_lists = []
+        online_hbas = nodedev.find_hbas("hba")
+        if not online_hbas:
+            logging.error("NO ONLINE HBAS!")
+        first_online_hba = online_hbas[0]
+        #create vhba based on the first online hba
+        old_vhbas = nodedev.find_hbas("vhba")
+        logging.debug("OLD OLD %s", old_vhbas)
```
Comment (lento-sun) [inline libvirt/tests/src/npiv/npiv_hostdev_passthrough.py]: s/OLD OLD/Original online vHBAs: /

File: libvirt/tests/src/npiv/npiv_hostdev_passthrough.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+    hostdev_xml.source = hostdev_xml.new_source(**source_args)
+    logging.info(hostdev_xml)
+    return hostdev_xml
+
+
+def find_scsi_luns(scsi_host):
+    """
+    Find available luns of specified scsi_host.
+
+    :param scsi_host: The scsi host name in foramt of "scsi_host#"
```
Comment (balamuruhans) [inline libvirt/tests/src/npiv/npiv_hostdev_passthrough.py]: typo %s/foramt/format

File: libvirt/tests/src/npiv/npiv_hostdev_passthrough.py [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+    :return: A dictionary contains all available fc luns
+    """
+    lun_dicts = []
+    tmp_list = []
+    scsi_number = scsi_host.replace("scsi_host", "")
+    cmd = "multipath -ll | grep '\- %s:' | grep 'ready running' |\
+           awk '{FS=\" \"}{for (f=1; f<=NF; f+=1) {if ($f ~ /%s:/)\
+           {print $f}}}'" % (scsi_number, scsi_number)
+    result = process.run(cmd, shell=True)
```
Comment (balamuruhans) [inline libvirt/tests/src/npiv/npiv_hostdev_passthrough.py]: It is better to have process.run in try block, If multipath is not installed you can handle and install it to continue with test.

File: libvirt/tests/src/npiv/npiv_hostdev_passthrough.py [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+        virsh_dargs = {'debug': True, 'ignore_status': True}
+        new_vhbas = []
+
+        if vm.is_dead():
+            vm.start()
+        session = vm.wait_for_login()
+        old_parts = libvirt.get_parts_list(session)
+        #find first online hba
+        online_hbas = []
+        vhba_lists = []
```
Comment (balamuruhans) [inline libvirt/tests/src/npiv/npiv_hostdev_passthrough.py]: where does this vhba_lists is used?

File: libvirt/tests/src/npiv/npiv_hostdev_passthrough.py (hunk 4 above)
Comment (lento-sun) [inline libvirt/tests/src/npiv/npiv_hostdev_passthrough.py]: modified. thx

File: libvirt/tests/src/npiv/npiv_hostdev_passthrough.py (hunk 3 above)
Comment (lento-sun) [inline libvirt/tests/src/npiv/npiv_hostdev_passthrough.py]: modified, thx

**Comment by lento-sun** [comment]
change code to import utils_npiv which is from https://github.com/avocado-framework/avocado-vt/pull/826

File: libvirt/tests/cfg/npiv/npiv_hostdev_passthrough.cfg [hunk 6]
```diff
@@ -0,0 +1,27 @@
+- npiv.hostdev_passthrough:
+    type = npiv_hostdev_passthrough
+    start_vm = "yes"
+#    vms = ""
+    scsi_wwnn = "20000000c99e2b80"
+    scsi_wwpn = "1000000000000001"
```
Comment (will-Do) [inline libvirt/tests/cfg/npiv/npiv_hostdev_passthrough.cfg]: Do we need to replace these values when running the case, if yes, you may set a more general value here, such as:
scsi_wwnn = " SCSI_WWNN_EXAMPLE"

Comment (balamuruhans) [inline libvirt/tests/cfg/npiv/npiv_hostdev_passthrough.cfg]: yes, right.

Comment (lento-sun) [inline libvirt/tests/cfg/npiv/npiv_hostdev_passthrough.cfg]: modified
and also change all avocado exceptions to test.fail/error/cancel

File: libvirt/tests/src/npiv/npiv_hostdev_passthrough.py [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+    4. login the vm and check the attached disk
+    5. detach-device the hostdev xml
+    6. login the vm to check the partitions
+    """
+    try:
+        status_error = params.get("status_error", "no")
+        vm_name = params.get("main_vm", "avocado-vt-vm1")
+        device_target = params.get("hostdev_disk_target", "hdb")
+        scsi_wwnn = params.get("scsi_wwnn", "ENTER.YOUR.WWNN")
+        scsi_wwpn = params.get("scsi_wwpn", "ENTER.YOUR.WWPN")
```
Comment (balamuruhans) [inline libvirt/tests/src/npiv/npiv_hostdev_passthrough.py]: we need to skip the tests if user fails to set respective params

File: libvirt/tests/src/npiv/npiv_hostdev_passthrough.py [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+        added_part = None
+        if target.startswith("vd"):
+            if added_parts[0].startswith("vd"):
+                added_part = added_parts[0]
+        elif target.startswith("hd"):
+            if added_parts[0].startswith("sd"):
+                added_part = added_parts[0]
+
+        if not added_part:
+            logging.error("Cann't see added partition in VM")
```
Comment (balamuruhans) [inline libvirt/tests/src/npiv/npiv_hostdev_passthrough.py]: `%s/cann't/can't`

File: libvirt/tests/src/npiv/npiv_hostdev_passthrough.py [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+            return False
+
+        cmd = ("fdisk -l /dev/{0} && mkfs.ext4 -F /dev/{0} && "
+               "mkdir -p test && mount /dev/{0} test && echo"
+               " teststring > test/testfile && umount test"
+               .format(added_part))
+        try:
+            s, o = session.cmd_status_output(cmd)
```
Comment (balamuruhans) [inline libvirt/tests/src/npiv/npiv_hostdev_passthrough.py]: let's use proper variable name here and elsewhere

File: libvirt/tests/src/npiv/npiv_hostdev_passthrough.py [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+        except Exception, e:
+            test.error("Error occurred when run cmd: fdisk")
+        logging.info("Check disk operation in VM:\n%s", o)
+        session.close()
+        if s != 0:
+            return False
+        return True
+    except (remote.LoginError, virt_vm.VMError, aexpect.ShellError), e:
+        logging.error(str(e))
```
Comment (balamuruhans) [inline libvirt/tests/src/npiv/npiv_hostdev_passthrough.py]: same here

File: libvirt/tests/src/npiv/npiv_hostdev_passthrough.py [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+        vm = env.get_vm(vm_name)
+        vmxml_backup = vm_xml.VMXML.new_from_inactive_dumpxml(vm_name)
+        virsh_dargs = {'debug': True, 'ignore_status': True}
+        new_vhbas = []
+
+        if vm.is_dead():
+            vm.start()
+        session = vm.wait_for_login()
+        old_parts = libvirt.get_parts_list(session)
+        #find first online hba
```
Comment (balamuruhans) [inline libvirt/tests/src/npiv/npiv_hostdev_passthrough.py]: space after `#` expected,
`# find first online hba`

File: libvirt/tests/src/npiv/npiv_hostdev_passthrough.py [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+        online_hbas = []
+        online_hbas = utils_npiv.find_hbas("hba")
+        if not online_hbas:
+            test.cancel("NO ONLINE HBAs!")
+        first_online_hba = online_hbas[0]
+        #create vhba based on the first online hba
```
Comment (balamuruhans) [inline libvirt/tests/src/npiv/npiv_hostdev_passthrough.py]: same here and elsewhere for comments

File: libvirt/tests/src/npiv/npiv_hostdev_passthrough.py [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+        logging.info("New hostdev xml as follow:")
+        logging.info(new_hostdev_xml)
+        new_hostdev_xml.xmltreefile.write()
+        if attach_method == "hot":
+            #attach-device the lun's hostdev xml to guest vm
+            result = virsh.attach_device(vm_name, new_hostdev_xml.xml)
+            status = result.exit_status
+            if status_error == "yes":
+                if status:
+                    test.fail("Attach device failed: %s" % result.stderr)
```
Comment (balamuruhans) [inline libvirt/tests/src/npiv/npiv_hostdev_passthrough.py]: `libvirt.check_exit_status(result)` can be used to reduce code

File: libvirt/tests/src/npiv/npiv_hostdev_passthrough.py [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+
+        #login vm and check the disk
+        check_result = check_in_vm(vm, device_target, old_parts)
+        if not check_result:
+            test.fail("check disk in vm failed")
+        result = virsh.detach_device(vm_name, new_hostdev_xml.xml)
+        status = result.exit_status
+        if status_error == "yes":
+            if status:
+                test.fail("Detach device failed: %s" % result.stderr)
```
Comment (balamuruhans) [inline libvirt/tests/src/npiv/npiv_hostdev_passthrough.py]: same here

File: libvirt/tests/src/npiv/npiv_hostdev_passthrough.py (hunk 8 above)
Comment (lento-sun) [reply libvirt/tests/src/npiv/npiv_hostdev_passthrough.py]: Update, thx

File: libvirt/tests/src/npiv/npiv_hostdev_passthrough.py (hunk 10 above)
Comment (lento-sun) [reply libvirt/tests/src/npiv/npiv_hostdev_passthrough.py]: update, thx

File: libvirt/tests/src/npiv/npiv_hostdev_passthrough.py (hunk 9 above)
Comment (lento-sun) [reply libvirt/tests/src/npiv/npiv_hostdev_passthrough.py]: update, thx

File: libvirt/tests/src/npiv/npiv_hostdev_passthrough.py (hunk 7 above)
Comment (lento-sun) [reply libvirt/tests/src/npiv/npiv_hostdev_passthrough.py]: update, thx

File: libvirt/tests/src/npiv/npiv_hostdev_passthrough.py (hunk 11 above)
Comment (lento-sun) [reply libvirt/tests/src/npiv/npiv_hostdev_passthrough.py]: update, thx

File: libvirt/tests/src/npiv/npiv_hostdev_passthrough.py (hunk 12 above)
Comment (lento-sun) [reply libvirt/tests/src/npiv/npiv_hostdev_passthrough.py]: update, thx

File: libvirt/tests/src/npiv/npiv_hostdev_passthrough.py (hunk 14 above)
Comment (lento-sun) [reply libvirt/tests/src/npiv/npiv_hostdev_passthrough.py]: update, thx

**Pull request merged**